### PR TITLE
Add error handling

### PR DESCRIPTION
**Pull request by mohit038**

### CODE REVIEW

**Comment by tobiasdiez** [comment]
In https://github.com/JabRef/JabRefOnline/issues/93, I've collected two nice resources on error handling in graphql. I still have to go through them and form an opinion about which approach to take. I'll try to do this evening. Do you have any suggestions or preferences?

**Comment by mohit038** [comment]
I prefer `Stage 5: Result Types`. It is a little hard to implement but has so many advantages.

**Comment by tobiasdiez** [comment]
I agree, this seems to be a nice strategy. I also liked some ideas of the "Stage 6" approach and of the zalando blog post, and updated the issue https://github.com/JabRef/JabRefOnline/issues/93 accordingly. 
One thing that I'm still a bit unsure about is the input field validation. I find it a bit exhausting to have a special type for each possible validation error (password to short, contains unsupported characters, is the same as old one, ...). I think, for theses cases where you normally have an input field and would then simply display the error message, these would be handled by the "InputValidationProblem". However, things that are more special would become a separate type, such as "TokenExpired" (because in this case you would like to alter the display on the frontend side considerably, i.e. redirect the user to the password reset page again). But on the other hand, the token is offically also only an input, so this feels a bit inconsistent. What is your opinion about this?

File: api/user/schema.graphql [hunk 1]
```diff
@@ -15,9 +15,34 @@ extend type Query {
...
   forgotPassword(email: EmailAddress!): Boolean
-  changePassword(token: String!, id: ID!, newPassword: String!): User
+  changePassword(
+    token: String!
+    id: ID!
+    newPassword: String!
+  ): ChangePasswordResponse
+}
+
+union ChangePasswordResponse = UserReturned | ExpiredTokenProblem
```
Comment (mohit038) [inline api/user/schema.graphql:27]: @tobiasdiez have a look at the graphql schema

File: api/user/auth.service.ts [hunk 2]
```diff
@@ -16,6 +16,27 @@ export interface AuthenticateReturn {
   message?: string
 }
 
+type UserReturned = {
```
Comment (tobiasdiez) [inline api/user/auth.service.ts:19]: Normally, these types are automatically generated using `yarn graphql:generate`

File: api/user/schema.graphql (hunk 1 above)
Comment (tobiasdiez) [reply api/user/schema.graphql:27]: Looks good to me, thanks! 

Not sure if the problem interfaces below need to be "types".

File: api/user/auth.service.ts (hunk 2 above)
Comment (mohit038) [reply api/user/auth.service.ts:19]: The `User` type used in auth service is of `'@prisma/client'` type, which has a different type than the `User` defined in graphql schema. 
error
`Type 'User' is missing the following properties from type 'User': documentsRaw, documents, groupsts(2739)`
So we have to overwrite the User type...

Comment (tobiasdiez) [reply api/user/auth.service.ts:19]: > The User type used in auth service is of '@prisma/client' type, which has a different type than the User defined in graphql schema.

That's right. However, in https://github.com/JabRef/JabRefOnline/blob/8a922d67a85d65d8c260e87de500c67b0452f31d/graphql.config.js#L17 we tell the graphql code generator that it should expect a prisma user to be returned whenever it sees a User type in the graphql schema. The transformation of a prisma user to a valid graphql type is handled by https://github.com/JabRef/JabRefOnline/blob/8a922d67a85d65d8c260e87de500c67b0452f31d/api/user/resolvers.ts#L103-L136

But probably I'm misunderstanding something...

Comment (tobiasdiez) [reply api/user/auth.service.ts]: @mohit038 I think I found a way to reuse the auto-generated types. At least it is working for me now. I've pushed the changes to this branch. Please have a look.

**Comment by tobiasdiez** [comment]
Is this indeed ready for review? Some of the code in the vue files is uncommented. I scrolled over the resolver code, and those changes look good.

Please also remove the mac store files (and add them to the git ignore).

**Comment by mohit038** [comment]
> Is this indeed ready for review? Some of the code in the vue files is uncommented. I scrolled over the resolver code, and those changes look good.
> 

No, I accidentally marked it ready for review. Sry

**Comment by tobiasdiez** [comment]
Do you have problems with the the client side codegen? This is currently configured at https://github.com/JabRef/JabRefOnline/blob/c0c60f88a7a561b15077d8cadfde037c5c495e2e/graphql.config.js#L32-L44

File: api/user/resolvers.ts [hunk 3]
```diff
@@ -48,58 +54,103 @@ export class Mutation {
     _root: Record<string, never>,
     { email, password }: MutationSignupArgs,
     context: Context
-  ): Promise<User> {
+  ): Promise<SignupPayload> {
     const newUser = await this.authService.createAccount(email, password)
-    await context.login(newUser)
+    if ('user' in newUser) void context.login(await newUser.user)
```
Comment (tobiasdiez) [inline api/user/resolvers.ts:59]: Please rename the variable to `newUserPayload` (since it's no longer only the new user)

File: components/Navbar.vue [hunk 4]
```diff
@@ -163,11 +163,12 @@
   </nav>
 </template>
 
+//
```
Comment (tobiasdiez) [inline components/Navbar.vue:166]: Please remove this

File: pages/change-password/_token.vue [hunk 5]
```diff
@@ -41,13 +42,27 @@ export default defineComponent({
     const password = ref('')
     const repeatPassword = ref('')
     gql`
-      mutation ChangePassword(
-        $token: String!
-        $id: ID!
-        $newPassword: String!
+      mutation ChangePasswordMutation(
+        $changePasswordToken: String!
```
Comment (tobiasdiez) [inline pages/change-password/_token.vue:46]: I think the "changePassword" prefix for the variables is unnecessary.

File: pages/change-password/_token.vue [hunk 6]
```diff
@@ -67,7 +82,7 @@ export default defineComponent({
     }))
     const router = useRouter()
     onDone(() => {
-      void router.push({ name: 'user-login' })
+      void router.push('../user/login')
```
Comment (tobiasdiez) [inline pages/change-password/_token.vue:85]: Is this change really necessary? (It's definitely unrelated to the actual purpose of this PR)

File: pages/user/forgot-password.vue [hunk 7]
```diff
@@ -30,25 +30,27 @@ import { defineComponent } from '@nuxtjs/composition-api'
...
 
 export default defineComponent({
   name: 'ForgotPassword',
   layout: 'bare',
   setup() {
     const email = ref('')
     gql`
-      mutation ForgotPassword($email: EmailAddress!) {
-        forgotPassword(email: $email)
+      mutation ForgotPasswordMutation($forgotPasswordEmail: EmailAddress!) {
```
Comment (tobiasdiez) [inline pages/user/forgot-password.vue:41]: here also, without prefix

File: pages/user/login.vue [hunk 8]
```diff
@@ -87,26 +87,43 @@ export default defineComponent({
     const otherError = ref('')
 
     gql`
-      mutation Login($email: EmailAddress!, $password: String!) {
-        login(email: $email, password: $password) {
-          id
+      mutation LoginMutation(
+        $loginEmail: EmailAddress!
```
Comment (tobiasdiez) [inline pages/user/login.vue:91]: without prefix

File: pages/user/login.vue [hunk 9]
```diff
@@ -87,26 +87,43 @@ export default defineComponent({
...
+        loginEmail: email.value,
+        loginPassword: password.value,
       },
     }))
     const router = useRouter()
     onDone((result) => {
       if (result.data?.login) {
-        currentUserVar(result.data.login)
+        currentUserVar(
+          result.data.login.__typename === 'UserReturned'
```
Comment (tobiasdiez) [inline pages/user/login.vue:123]: I think this check should be moved to the outer if check, and in the else statement `otherError.value` should be set to the message of the problem.

File: api/user/schema.graphql [hunk 10]
```diff
@@ -13,11 +13,46 @@ extend type Query {
...
+  user: User!
+}
+
+type InputValidationProblem {
+  problems: [InputFieldValidationProblem]!
+}
+
+type ExpiredTokenProblem {
+  problems: [InputFieldValidationProblem]!
+}
```
Comment (tobiasdiez) [inline api/user/schema.graphql:51]:
```suggestion
type ExpiredTokenProblem {
  message: String!
}
```
since the name of the problem already conveys that it is about a certain problem with the token.

File: api/user/schema.graphql [hunk 11]
```diff
@@ -13,11 +13,46 @@ extend type Query {
...
+  changePassword(
+    token: String!
+    id: ID!
+    newPassword: String!
+  ): ChangePasswordPayload
+}
+
+union LoginPayload = UserReturned | InputValidationProblem
+
+union ChangePasswordPayload = UserReturned | ExpiredTokenProblem
```
Comment (tobiasdiez) [inline api/user/schema.graphql:29]:
```suggestion
union ChangePasswordPayload = UserReturned | InputValidationProblem | ExpiredTokenProblem
```
if the new password is invalid?

File: pages/change-password/_token.vue [hunk 12]
```diff
@@ -22,6 +22,7 @@
   </div>
 </template>
 
+//
```
Comment (tobiasdiez) [inline pages/change-password/_token.vue:25]: also remove

**Review by tobiasdiez — CHANGES_REQUESTED** [review]
Looks really nice! Only the usual nitpicking left... 🦊

**Comment by tobiasdiez** [comment]
Thanks again!
